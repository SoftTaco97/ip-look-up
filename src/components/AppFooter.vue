<template>
  <v-container fluid>
    <v-row justify="center">
      <h3>Jared Martinez | {{ new Date().getFullYear() }} &copy; </h3>
    </v-row>
  </v-container>
</template>
