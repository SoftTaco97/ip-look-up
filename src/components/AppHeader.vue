<template>
  <v-container fluid>
    <v-row justify="center">
      <v-col cols="12">
        <h2>IP Address Look Up</h2>
      </v-col>
    </v-row>
    <search-form
      :search="ipAddress"
      @update-search="updateIpAddress"
      @submit="requestIpData"
    />
  </v-container>
</template>

<script>
import SearchForm from './SearchForm.vue';

export default {
  name: 'AppHeader',
  props: {
    ipAddress: {
      type: String,
      required: true,
    },
  },
  methods: {
    updateIpAddress(val) {
      this.$emit('update-search', val);
    },
    requestIpData() {
      this.$emit('submit');
    },
  },
  components: { SearchForm },
};
</script>
