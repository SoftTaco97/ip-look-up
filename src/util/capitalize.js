/**
 * Function for capitalizing a string
 * @param { String } str
 * @return { String }
 */
export default (str) => str.charAt(0).toUpperCase() + str.slice(1);
