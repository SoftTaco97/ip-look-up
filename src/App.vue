<template>
  <v-app>
    <header>
      <AppHeader
        :userSearch="userSearch"
        :loading="loading"
        @update-search="updateSearch"
        @submit="processIp"
      />
    </header>
    <v-main>
      <v-container fluid>
        <v-row justify="center" id="resultsArea">
          <v-col cols="8" md="6" id="ipResults">
            <v-sheet rounded elevation="2">
              <v-container fluid>
                <v-row justify="center" no-gutters>
                  <v-col
                    v-for="({ title, value }, index) in ipData"
                    :key="index"
                    cols="12"
                    md="6"
                    lg="3"
                    class="text-center text-md-left"
                  >
                    <small>{{ title }}</small>
                    <h3>{{ value }}</h3>
                  </v-col>
                  <v-col v-if="ipData.length <= 0">
                    Loading...
                  </v-col>
                </v-row>
              </v-container>
            </v-sheet>
          </v-col>
          <v-col cols="12" id="map"></v-col>
        </v-row>
      </v-container>
    </v-main>
    <v-footer
      app
      padless
      color="indigo accent-3"
    >
      <AppFooter />
    </v-footer>
    <v-snackbar
      v-model="error"
      top
      right
      color="error"
    >
      <v-icon class="mr-2">mdi-cloud-alert</v-icon>
      Whoops! Can't seem to request that IP
    </v-snackbar>
  </v-app>
</template>

<script>
import axios from 'axios';
import createMap from './util/map';
import capitalize from './util/capitalize';
import AppFooter from './components/AppFooter.vue';
import AppHeader from './components/AppHeader.vue';

export default {
  name: 'App',
  data() {
    return {
      userSearch: '',
      ipData: [],
      updateMap: null,
      error: false,
      loading: false,
    };
  },
  mounted() {
    // Load the data
    this.processIp();
  },
  methods: {
    mapTitle(value) {
      const titles = {
        isp: 'Internet Service Provider',
        ip: 'IP Address',
      };
      return titles[value] || capitalize(value);
    },
    updateSearch(search) {
      this.userSearch = search;
    },
    createIpData({ location, ip, isp }) {
      // Create a readable location
      const locationVal = `${location.city}, ${location.region} ${location.postalCode}`;
      // Create an object of data we need to showcase
      const dataObj = {
        ip,
        location: locationVal,
        timezone: location.timezone,
        isp,
      };
      // Loop over the data needed to showcase and return it
      return Object.entries(dataObj).map(([key, value]) => ({
        title: this.mapTitle(key),
        value,
      }));
    },
    setMap({ lat, lng } = {}) {
      // Set the location of the map based on if the updateMap property is set
      if (!this.updateMap) {
        this.updateMap = createMap(lat, lng);
      } else {
        this.updateMap(lat, lng);
      }
    },
    async processIp() {
      this.loading = true;
      try {
        const { location, isp, ip } = await this.requestIpData() || {};
        if (location && isp && ip) {
          this.ipData = this.createIpData({ location, isp, ip });
          // Set the ip address variable if it is not set
          if (!this.userSearch) this.updateSearch(ip);

          // Update the map
          this.setMap(location);
        }
      } catch (e) {
        this.error = true;
      } finally {
        this.loading = false;
      }
    },
    async requestIpData() {
      const url = `https://geo.ipify.org/api/v1?apiKey=${process.env.VUE_APP_API_KEY}&domain=${this.userSearch}`;
      const { status, data } = await axios.get(url);
      if (status === 200 && data) return data;
      throw new Error('Bad information returned');
    },
  },
  components: { AppHeader, AppFooter },
};
</script>

<style>
  header {
    background-image: url('assets/pattern-bg.png');
    background-size: cover;
    background-position: center center;
  }

  header h2, footer h3 {
    color: #ffffff;
    font-family: 'Roboto', sans-serif;
    font-weight: normal;
  }

  header h2 {
    text-align: center;
  }

  #ipTracker {
    padding-top: 0;
  }
  #resultsArea {
    position: relative;
  }
  #ipResults, #map {
    position: absolute;
  }
  #ipResults {
    z-index: 9;
    margin-top: -3.5em;
  }
  #map {
    z-index: 0;
    margin-top: -1em;
    height: calc(100vh - 22.5em);
    border-radius: 5px;
  }

  /* Desktop styling */
  @media screen and (min-width: 1025px) {
    .v-sheet > .container > .row > .col-md-6 {
      padding: 0 .5em;
    }
    .v-sheet > .container > .row > .col-md-6:not(:last-of-type) {
      border-right: 1px solid lightgray;
    }
  }
</style>
