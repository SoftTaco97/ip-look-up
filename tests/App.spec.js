import axios from 'axios';
import createMap from '../src/util/map';
import capitalize from '../src/util/capitalize';
import createComponentMounter from './componentMounter';
import AppHeader from '../src/components/AppHeader.vue';
import AppFooter from '../src/components/AppFooter.vue';
import App from '../src/App.vue';

jest.mock('axios', () => ({
  get: jest.fn().mockResolvedValue({
    status: 200,
    data: [],
  }),
}));
jest.mock('../src/util/map', () => jest.fn().mockReturnValue(jest.fn()));
jest.mock('../src/util/capitalize', () => jest.fn().mockImplementation((str) => str));

describe('App.vue unit tests', () => {
  const mountComponent = createComponentMounter(App);
  let wrapper;

  beforeEach(() => {
    wrapper = mountComponent();
  });

  afterEach(() => {
    wrapper.destroy();
    jest.clearAllMocks();
  });

  describe('rendering unit tests', () => {
    it('Should match the snapshot', () => {
      // ! - Deep mount for snapshots
      wrapper = mountComponent({}, false);
      expect(wrapper).toMatchSnapshot();
    });
    it('Should render properly', () => {
      expect(wrapper.exists()).toEqual(true);
    });
    it('Should have a header and footer ', () => {
      expect(wrapper.findComponent(AppHeader).exists()).toEqual(true);
      expect(wrapper.findComponent(AppFooter).exists()).toEqual(true);
    });
    it('Should display a loading message when there is no ip data', () => {
      expect(wrapper.text()).toEqual(expect.stringContaining('Loading...'));
    });
    it('Should display the ip data', async () => {
      const mockIpData = [
        {
          title: 'IP Address',
          value: '111.111.111.111',
        },
        {
          title: 'Location',
          value: 'Some City, AA, 12345',
        },
        {
          title: 'Timezone',
          value: '-8:00',
        },
        {
          title: 'Internet Service Provider',
          value: 'Comcast sux',
        },
      ];
      await wrapper.setData({ ipData: mockIpData });
      const wrapperText = wrapper.text();
      // Loop through the data and check for it
      mockIpData.forEach(({ title, value }) => {
        expect(wrapperText).toEqual(expect.stringContaining(title));
        expect(wrapperText).toEqual(expect.stringContaining(value));
      });
    });
    it('Should show an error message when an error is encountered', async () => {
      await wrapper.setData({ error: true });
      expect(wrapper.text()).toEqual(expect.stringContaining('Whoops! Can\'t seem to request that IP'));
    });
  });
  describe('method unit tests', () => {
    describe('mapTitle unit tests', () => {
      const mockTitles = { test: 'This is a tes title!' };

      beforeEach(() => wrapper.setData({ titles: mockTitles }));

      it('Should use the value in data.titles when there is one present', () => {
        expect(wrapper.vm.mapTitle('test')).toEqual(mockTitles.test);
      });
      it('Should capitalize the title provided when it is not in data.titles', () => {
        const mockTitle = 'someOtherTitle';
        wrapper.vm.mapTitle(mockTitle);
        expect(capitalize).toHaveBeenCalledTimes(1);
        expect(capitalize).toHaveBeenCalledWith(mockTitle);
      });
    });
    // describe('updateSearch unit tests', () => {});
    // describe('createIpData unit tests', () => {});
    // describe('setMap unit tests', () => {});
    // describe('processIp unit tests', () => {});
    // describe('requestIpData unit tests', () => {});
  });
});
