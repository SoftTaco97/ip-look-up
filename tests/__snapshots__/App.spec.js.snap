// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue unit tests rendering unit tests Should match the snapshot 1`] = `
<div data-app="true" class="v-application v-application--is-ltr theme--light" id="app">
  <div class="v-application--wrap">
    <header>
      <div class="container container--fluid">
        <div class="row justify-center">
          <div class="col col-12">
            <h2>IP Address Look Up</h2>
          </div>
        </div>
        <div class="row justify-center">
          <div class="col-md-6 col-lg-4 col-12">
            <form novalidate="novalidate" class="v-form">
              <div class="row">
                <div class="col-md-10 col-12">
                  <div class="v-input theme--dark v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <fieldset aria-hidden="true">
                          <legend style="width: 0px;"><span>​</span></legend>
                        </fieldset>
                        <div class="v-text-field__slot"><label for="input-15" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">IP Address</label><input id="input-15" placeholder="Enter IP or Domain Here.." type="text"></div>
                        <div class="v-input__append-inner">
                          <div class="v-input__icon v-input__icon--clear"><button disabled="disabled" type="button" aria-label="clear icon" class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--dark"></button></div>
                        </div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--dark">
                          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="col-md-2 col-12"><button type="submit" disabled="disabled" class="white--text v-btn v-btn--block v-btn--contained v-btn--disabled theme--light v-size--large" aria-label="search"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-magnify theme--light"></i></span></button></div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </header>
    <main class="v-main" style="padding: 0px 0px 0px 0px;">
      <div class="v-main__wrap">
        <div class="container container--fluid">
          <div id="resultsArea" class="row justify-center">
            <div id="ipResults" class="col-md-6 col-8">
              <div class="v-sheet theme--light elevation-2 rounded">
                <div class="container container--fluid">
                  <div class="row no-gutters justify-center">
                    <div class="col">
                      Loading...
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div id="map" class="col col-12"></div>
          </div>
        </div>
      </div>
    </main>
    <footer class="v-footer v-sheet theme--light v-footer--fixed v-footer--padless indigo accent-3" style="left: 0px; right: 0px; bottom: 0px;">
      <div class="container container--fluid">
        <div class="row justify-center">
          <h3>Jared Martinez | 2020 © </h3>
        </div>
      </div>
    </footer>
    <div class="v-snack v-snack--has-background v-snack--right v-snack--top" style="padding-bottom: 0px; padding-top: 0px;">
      <transition-stub name="v-snack-transition">
        <div class="v-snack__wrapper v-sheet theme--dark error" style="display: none;">
          <div role="status" aria-live="polite" class="v-snack__content"><i aria-hidden="true" class="v-icon notranslate mr-2 mdi mdi-cloud-alert theme--dark"></i>
            Whoops! Can't seem to request that IP
          </div>
          <div class="v-snack__action "></div>
        </div>
      </transition-stub>
    </div>
  </div>
</div>
`;
